### PR TITLE
Feature/integrate add fcm token

### PR DESCRIPTION
**Pull request by sieuhuflit**
## Overview

Integrate with backend **addFcmToken & removeFcmToken**
- [x] Trigger call `addFcmToken` when open app & after login successfully 
- [x] Trigger call `removeFcmToken` when logout

### CODE REVIEW

File: lib/core/service/auth/auth_service.dart [hunk 1]
```diff
@@ -4,28 +4,33 @@ import 'package:app/core/domain/auth/entities/auth_session.dart';
...
 
   Future<Either<Failure, bool>> login() async {
     var res = await appOAuth.login();
     return res.fold(
       (l) => Left(Failure()),
       (success) {
-        if (success) return const Right(true);
-
+        if (success) {
+          firebaseService.addFcmToken();
```
Comment (mthinh) [inline lib/core/service/auth/auth_service.dart:24]: I recommend not putting fcmToken logic inside auth service, 
In firebase service, let create a listener on `getIt<AppOAuth>().tokenStateStream, 
When ever the tokenState changed to valid, then we should addFcmToken to our server

File: lib/app.dart [hunk 2]
```diff
@@ -70,6 +70,7 @@ class _LemonadeAppViewState extends State<LemonadeApp> {
     WidgetsBinding.instance.addPostFrameCallback((_) async {
       FirebaseService.setContext(context);
     });
+    FirebaseService().addFcmToken();
```
Comment (mthinh) [inline lib/app.dart:73]: Wonder why we addFcmToken here ?

File: lib/core/service/auth/auth_service.dart [hunk 3]
```diff
@@ -4,28 +4,33 @@ import 'package:app/core/domain/auth/entities/auth_session.dart';
...
+          return const Right(true);
+        }
         return Left(Failure());
       },
     );
   }
 
   Future<Either<Failure, bool>> logout() async {
+    firebaseService.removeFcmToken();
```
Comment (mthinh) [inline lib/core/service/auth/auth_service.dart:33]: Due to required server-accessToken to remove fcmToken, it's ok to put this logic here

File: lib/core/service/firebase/firebase_service.dart [hunk 4]
```diff
@@ -156,4 +163,38 @@ class FirebaseService {
     FirebaseMessaging.onMessageOpenedApp
         .listen(_firebaseMessagingBackgroundHandler);
   }
+
+  void addFcmToken() async  {
+    var tokenState = await getIt<AppOauth>().tokenState;
```
Comment (mthinh) [inline lib/core/service/firebase/firebase_service.dart:168]: we can use `tokenStateStream` to listen of tokenState changed and decide to addFcmToken or not

Comment (sieuhuflit) [reply lib/core/service/firebase/firebase_service.dart]: Yep, sound good 👍

File: lib/core/service/auth/auth_service.dart (hunk 1 above)
Comment (sieuhuflit) [reply lib/core/service/auth/auth_service.dart]: Yep, let me adjust

File: lib/main_production.dart [hunk 5]
```diff
@@ -11,14 +11,12 @@ void main() async {
...
-  final firebaseService = FirebaseService();
-  await firebaseService.initialize();
-  final token = await firebaseService.getToken();
-  debugPrint('FCM Token: $token');
-  
   registerModule();
   
   await getIt<AppOauth>().init();
+  
+  final firebaseService = FirebaseService();
```
Comment (mthinh) [inline lib/main_production.dart:18]: because you already registered firebaseService as LazySingleton, just need to `getIt<FirebaseService>().initialize()`

Comment (sieuhuflit) [reply lib/main_production.dart]: Updated 👍

**Comment by mthinh** [comment]
nice work 👍

**Review by mthinh — APPROVED** [review]

**Pull request merged**